### DISCUSSION TITLE
Windows WSL support

### DISCUSSION THREAD

**Discussion by poetaman** [poetaman/arttime#24] [General]
This is a place to discuss how to make arttime work with all its bells and whistles on Windows WSL.

**Comment by Null-Sector** [comment]
Just curious on what's the easiest way to get it up and running on any Windows WSL and/or PowerShell. I'm using Windows Terminal.

**Reply by poetaman** [reply]
@Null-Sector Unfortunately I don't have access to a Windows machine at the moment to be of help. In my understanding, ``zsh`` (>=5.7)  should be sufficient to get it up and running. But it would be without sounds/notifications as I still don't know what command could be used from bash shell in WSL to push a Windows desktop notification and generate sound. Check the dependency section on ``arttime`` wiki for https://github.com/poetaman/arttime/wiki#windows-wsl. Arttime is looking for volunteers on that front! Not sure what terminal you use, I would go for something that displays Unicode glyphs. It would additionally be good to have ``less`` installed.

To check if some command is installed on WSL, do ``command -v command_name``; for instance ``command -v zsh``, ``command -v less``.

Once you have ``zsh`` installed, just ``cd`` to arttime's repository (you don't even need to install it), and run ``./bin/arttime``. Additionally, if ``arttime`` runs like that, run ``./install.sh`` and please post the output here in a .zip file. That would be of great help...

**Reply by Null-Sector** [reply]
I tried `zsh` in my Kali WSL 2.0 install, it did nothing but I'm also not the best with Linux terminals, I'm far more comfy inside PowerShell. I'll do some prodding around but do feel free to DM me or get back to me when you can get to a Windows machine. I do appreciate the swift reply and attempt to help regardless!

**Reply by poetaman** [reply]
@Null-Sector I didn't know Kali Linux supported WSL. Their website seems to suggest 3 different modes for running WSL. Did you try it in a Linux terminal inside Linux's Desktop Environment. Or inside Windows's Desktop Environment with WSL access done via just a shell? You are welcome! Here's the link: https://www.kali.org/docs/wsl/win-kex/

**Reply by Null-Sector** [reply]
Oh yeah, it works pretty flawlessly [only know, if you do it, at first it's a hyper barebones Kali install and you have to manually install your pen testing tools and stuffs.] I've used KeX but yo that's a good point. I was using Kali in the Windows Terminal so yeah under Windows WSL directly. I should try it with KeX.

**Reply by poetaman** [reply]
"Oh yeah, it works pretty flawlessly". I guess you mean inside a Linux Desktop Environment of some Linux VM running on Windows.

**Reply by Null-Sector** [reply]
Well seeing as WSL has objectively better performance due to Hyper-V and me not needing most of a Linux GUI unless it's some weird Kali thing I gotta boot up KeX for. Yeah works really well for me and many other people lol.

**Reply by poetaman** [reply]
@Null-Sector From https://github.com/poetaman/arttime/issues/53, it looks like arttime now works well under WSL with native Windows notifications and sounds. The dependencies are: ``zsh`` (>=5.7), [wsl-notify-send](https://github.com/stuartleeks/wsl-notify-send), and ``less``. Can you please pull the main branch, ``cd`` to arttime repo, and try ``./bin/arttime`` or ``TERM=xterm-256color ./bin/arttime`` after installing the dependencies? If it doesn't work, or you have some issues, feel free to drop a comment here or open a new issue.

**Comment by panki27** [comment]
I've just tried `arttime` on my WSL2 running Debian 11 inside Microsoft Terminal.
On a glance, it seems to work *mostly fine*, but it returns 2 errors on starting:

```
echoti:539: no such terminfo capability: rmam
echoti:541: no such terminfo capability: smam
```
I haven't tested the goal stuff yet.
And I can't get the tarot cards to show up whitout missing symbols (`🭸`), no matter which font I pick in MS Terminal (even C64 Pro Mono doesn't show them). Any ideas?

**Reply by poetaman** [reply]
@panki27 Your terminfo database is either stale or it does not represent your terminal well. I have seen that base Linux/Unix installs can still carry old terminfo databases that don't really speak truth about the terminal. For instance, missing ``smam``/``rmam`` pair means that your terminal lacks line wrapping/truncation capability, which is not true for any terminal emulator of our times... Generally one can install latest terminfo databases from a package in one's Linux/Unix distribution.

1) What does ``echo $TERM`` return?

2) What if you try launching arttime like this: ``TERM=xterm-256color <path_to_arttime>/arttime`` or assuming it is already installed: ``TERM=xterm-256color arttime``

For missing glyph, give the font build in this comment a try: https://github.com/shytikov/pragmasevka/issues/1#issuecomment-1328643230. If it does not show correctly either, could you try another terminal? Maybe there is a bug in Windows terminal. A terminal you try shouldn't be using vte, check my note on terminals in https://github.com/poetaman/arttime/wiki#terminals, not sure if that even applies for terminals in Windows.

Regarding desktop notifications and sounds: I still don't know if Windows has a command available under WSL* that would show native desktop notifications and play native desktop sounds.

3) Does ``notify-send`` command exist and generates desktop notification. Example: ``notify-send "Hello World"``? If command is not found, you would need to install package ``libnotify-bin`` that provides [notify-send](https://manpages.debian.org/testing/libnotify-bin/notify-send.1.en.html).

4) Check if sounds can be played from WSL by installing package ``vorbis-tools`` and running some file with [ogg123](https://manpages.debian.org/jessie/vorbis-tools/ogg123.1.en.html) command. It requires sounds in .ogg format (which freedesktop provides). If not, check if ``paplay`` command exists under WSL and plays sounds?

**Reply by panki27** [reply]
@poetaman thanks for your reply.

1. My `$TERM` is set to `tmux-256color`. 
2. Running it with `xterm-256color` stops the warnings I posted from appearing! 👍
3. `notify-send` runs without issue but doesn't generate a desktop notification (or anything notable).
4. I don't have any audio system installed (this is terminal only) so I doubt this is going to work.
However, you can run any Windows binary from within WSL. I've tested [this](https://superuser.com/a/1533515/680443) solution, and I can confirm it works. This could be stored as a script (`powershell.exe -file alert.ps1`) or potentially ran within a `-ScriptBlock` argument for `powershell.exe`.

I've given the font a shot, thanks for that. The tarot cards look a lot better, almost right, but not quite - I've attached an image.

![tarot](https://user-images.githubusercontent.com/18188892/209350460-66eee6a5-a344-4a0c-9f40-49ee341cc642.png)

**Reply by poetaman** [reply]
@panki27 Yes, new tmux-256colors should have rmam/smam. I would recommend updating your terminfo database, or just use xterm-256color... This is not the only app you would come across that uses rmam/smam capabilities. The art in image looks so off. Does this also happen with other art files which have unicode characters like ``monalisa9``? If it is not centered then there is a locale issue. Check my first bullet point in [arttime troubleshooting](https://github.com/poetaman/arttime#troubleshooting). Am attaching the text here for your reference:

># Troubleshooting
>- Some art does not render correctly?
>  - If the problem is that some arts are centered while others are not, it is because your locale does not have UTF-8 in it. For instance look at the docker [command](https://github.com/poetaman/arttime#try-arttime-in-docker) on this same page. You would require a setting like ``LC_ALL=C.UTF-8``. Unfortunately the same string might not work as there is no standard naming convention (it could be named ``en_US.UTF-8`` instead of ``C.UTF-8`` among other possibilities). Web search is the only way to figure the exact name of UTF-8 locale.

Try something like this?
```
TERM=xterm-256color LC_ALL=C.UTF-8 arttime
```

If that doesn't work, then can you try logging into WSL from another terminal application instead of Windows terminal?

Thanks for the pointer to powershell script. I don't like the idea of saving the code in a file and the documentation of powershell's ``-Command`` states that _"You can specify a script block only when running PowerShell in PowerShell."_ The same restriction might not apply to passing commands via stdin. PRs are welcome if you could figure a way of passing those commands via stdin or string when running from zsh/bash as the outer shell. I cannot test it on AppleSilicon Mac as it does not support nested virtualization... As a sidenote I uninstalled powershell from macOS within 30 min of installing it. It was generating activity on-and-off while shell was idling at the prompt. A quick search confirms it's a usual Microsoft spywear :D. Check this [link](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_telemetry?view=powershell-7.3), a hidden in docs "opt-out" instead of getting permission from user to opt-in, classic Microsoft tactic.

**Reply by panki27** [reply]
@poetaman thanks for your input, I've switched to `xterm-256color` in my `.tmux.conf`.

While my `LC_ALL` is unset, setting it to `C.UTF-8` has no effect on the output. I checked the correctness of `C.UTF-8` with `locale -a` - it's listed, but `en_US.utf8` is listed also. The latter doesn't produce a change either.

I've installed WezTerminal and set up the correct font (the one you provided), it unfortunately looks the same.
I've hit "random art" quite a few times now and checked `monalisa9` - everything looks perfect, it's just the tarot cards that are having issues.

Also, I've tried all this outside `tmux` aswell to rule that out.

And regarding your last point: yeah there is no point at all in installing powershell on a *nix system.
You can easily [pass arguments to powershell scripts on the commandline](https://www.delftstack.com/howto/powershell/command-line-arguments-in-powershell/). I might look into creating a PR sometime, but I'm not good at shell scripting.

**Reply by poetaman** [reply]
@panki27 Ok, thanks for testing. Not sure if this is because terminals on Windows rely on some library that does not like [Symbols for legacy computing](https://en.wikipedia.org/wiki/Symbols_for_Legacy_Computing), or zsh or some unix command I am using.

Let's try a couple more simple commands to narrow down the cause:

* First ``cd`` to directory [share/arttime/textart](https://github.com/poetaman/arttime/tree/main/share/arttime/textart) in 1) arttime's repository or 2) it's install directory:

```
# Option-1
cd <path_to_arttime_repository>/share/arttime/textart
# Option-2
cd ~/.local/share/arttime/textart
```

* Run these commands on file ``tarot-0major_arcana13``. If the output does or does not look like in the screenshot below, knowing that would be one more bit of useful debug information (that I could feed into my AI debug bot, jk).

```
cat tarot-0major_arcana13
pr -to tarot-0major_arcana13
```

<img width="722" alt="Screenshot 2022-12-24 at 4 33 22 PM" src="https://user-images.githubusercontent.com/71736629/209454003-6b71a91e-31f4-4a22-957d-1ec89f28a593.png">

**Reply by panki27** [reply]
@poetaman Thanks a lot for taking care of this :) See attachments for results (switched to Solarized for better visibility).

![tarot1](https://user-images.githubusercontent.com/18188892/209558796-ed21ac3f-8624-46a9-ade5-a8ed0fc8e9ab.png)
![tarot2](https://user-images.githubusercontent.com/18188892/209558803-a239c47f-00c1-4ba6-81bb-16d11781a1d0.png)

I've double checked that my `xterm-256colors` is right for the MS terminal, and [it appears to be.](https://superuser.com/a/1690394/680443)

I've taken a look at the bytes in the first line with `xxd`.
```
head -n 2 tarot-0major_arcana13 | tail -n 1 | xxd
```

What I can see are ANSI escape sequences and some bytes I don't know what to do with (the first four being `e295ad1b`, right after the first color code).
The escape sequences seem to be bg/fg color only (`\033[38;2;255;255;255;48;2;0;0;0m`), and they work correctly in `echo`.

**Reply by panki27** [reply]
... I just found [this](https://github.com/microsoft/terminal/issues/3177), which might also be related.

**Reply by poetaman** [reply]
@panki27 ``e295ad1b`` is expected and correct. ``e295ad`` is the UTF-8 hex code for first character ``╭`` ([U+256D](https://www.compart.com/en/unicode/U+256D)), and ``1b`` is the hex code for escape character ``\033`` (in this case the escape character for the next escape sequence). Microsoft terminal issue you point to seems unrelated for 3 reasons: 1) art files does not have any ``\e[K``, nor does arttime use ``\e[K``, 2) same problem is seen on WezTerm in Windows, 3) the first line with escape sequence looks off, which cannot be affected by any problem wrapping.

@wez Would you know why certain fg/bg color escape sequences don't work right on WezTerm in Windows but does on WezTerm in macOS and Linux? Please see the screenshots in this thread above. Same example file under question is attached here: [termtest.txt](https://github.com/poetaman/arttime/files/10304912/termtest.txt). Run ``cat termtest.txt`` in WezTerm on Windows, and on macOS. The text art will look good on macOS, and not so much on Windows. Let us know if it is worth filing a bug on WezTerm... Multiple terminals on Windows seem affected so we don't know where to begin. It would be good to have you see it as you have a terminal that runs on multiple platforms.

**Reply by wez** [reply]
I only skimmed this issue, but it made me think of the set of related ConPTY issues below, where certain escape sequences can confuse the ConPTY layer and produce surprising results:

* https://github.com/microsoft/terminal/issues/13710#issuecomment-1230729911 (the underlying main tracking issue)
* https://github.com/nushell/nushell/issues/5585#issuecomment-1230417496 which surfaces in nushell
* https://github.com/wez/wezterm/issues/2442#issuecomment-1221670772 which demonstrates some of the differences

Separately, this issue discusses a problem in ConPTY where it respects one form of color escape but not another (colons vs. semicolons):
* https://github.com/wez/wezterm/issues/2723

If I had to guess, I'd suggest that you split `^[[38;2;255;255;255;48;2;0;0;0m` into two separate escapes: `^[[38;2;255;255;255m` and `^[[48;2;0;0;0m` and see if ConPTY is happier about it.

**Reply by poetaman** [reply]
@wez Thanks a lot! @panki27 please give [termtest1.txt](https://github.com/poetaman/arttime/files/10305869/termtest1.txt) a try with same commands. If that works then I can break the escape sequences in art files to be separate like wez suggests. If not, then we have a new bug for ConPTY.

**Reply by panki27** [reply]
@poetaman I've tried it, but unfortunately no change. Thanks for the effort!

**Reply by poetaman** [reply]
@panki27 Ok thanks! Let's do this one last test. In the following file I have removed all 24-bit background color sequences like ``^[[48;2;0;0;0m``. If the form of art is not messed up like in the presence of 24-bit background color sequences, then we know it happens only when there are 24-bit background color escape sequences: [termtest2.txt](https://github.com/poetaman/arttime/files/10311536/termtest2.txt). Here's the screenshot without any background color:

<img width="357" alt="Screenshot 2022-12-27 at 5 43 07 PM" src="https://user-images.githubusercontent.com/71736629/209744131-c9a42bda-2bbb-4543-b49f-4a4e0804ced7.png">

**Reply by panki27** [reply]
@poetaman Tested, no effect, thanks for the effort.
Got a friend of mine to test in his WSL aswell, same behaviour.

**Reply by wez** [reply]
I took termtest1.txt, and reduced it to the first line with graphics and color, and trimmed it down to the first couple of cells, then used `wezterm record` to capture what conpty outputs when running `type termtest1.txt`. 

Here's my modified input:

```
^[[38;2;255;255;255m^[[48;2;0;0;0m╭^[[38;2;255;255;255m^[[48;2;0;0;0m🭹 ^[[38;2;255;255;255m^[[48;2;0;0;0m🭹 ^[ ^[[0m
```

 The significant line of output from the recording is:

```
"\u001b[38;2;255;255;255m\u001b[48;2;0;0;0m\r\n╭🭹\u001b[1C🭹\u001b[38;5;2m\u001b[49m\r\n07:02 "
```

You can see that ConPTY has re-ordered the color escapes around the newline(!) and inserted `ESC [ 1C` and then altered the trailing escapes significantly to `38;5;2` and `49`

Poking around the issue tracker, I think this is the root cause bug:

* https://github.com/microsoft/terminal/issues/8698

**Reply by wychwitch** [reply]
I'm so glad I found this, because after discovering the tarot mode I could not figure out why they weren't rendering correctly (kept installing different nerdfonts, convinced I messed something up)

I appreciate you looking into it @wez ! i guess we gotta wait for Conpty to be fixed!

**Reply by poetaman** [reply]
@wychwitch Good you found this! I wish more GitHub repos enabled discussions. The Windows WSL support is stalled at the moment as I do not have access to a windows machine that can run WSL... It might take some time before arttime gets desktop notifications and sounds on WSL.

**Reply by wychwitch** [reply]
> @wychwitch It might take some time before arttime gets desktop notifications and sounds on WSL.

Honestly I didn't even realize it was even possible to have notifications and sounds in WSL! I just accepted that as a limitation Id live with hahaha

I use WSL (specifically Arch) so if there's any help I could give for tests and such let me know! Tho I get it if that's not really feasible with the nature of development lmao. I'm not familiar with ZSH development otherwise I would make some PRS myself!

**Reply by poetaman** [reply]
@wychwitch The change needed would be simple. Here are some experiments needed:

There is a Linux command notify-send clone for WSL here: [wsl-notify-send](https://github.com/stuartleeks/wsl-notify-send), and a related [article](https://stuartleeks.com/posts/wsl-github-cli-windows-notifications-part-1/). The first thing to do is check if that really works. Also, does it make some default sound? Please try the following commands, and post a screenshot of notification.

```
# First add notify-send function to bash or zsh as described on README.md of wsl-notify-send
notify-send -u critical "ARTTIME" "Sat May 06, 2023 12:16:55AM PDT\rgoal: 1m\rart: buttefly\rID: 23342"
# Try it without the wrapper too
wsl-notify-send.exe --category $WSL_DISTRO_NAME -u critical "ARTTIME" "Sat May 06, 2023 12:16:55AM PDT\rgoal: 1m\rart: buttefly\rID: 23342"
```

If that works, then the next step would be for arttime to detect it is running in WSL. Can you provide the output of these commands: ``$ uname``, ``$ uname -a``, ``uname -s``, and ``uname -r``?

An alternative to wsl-notify-send is [BurntToast](https://github.com/Windos/BurntToast). Though it's not necessary to test this if wsl-notify-send works. It requires powershell, so the question then would be... is powershell.exe available from Linux running under WSL by default or is it something one needs to install from [AUR](https://aur.archlinux.org/packages/powershell)?

zsh is pretty compatible with bash syntax for basic operations. But I can do the edits in a branch for you to test based on the results of experiments mentioned above.

Thanks for your willingness and offer to collaborate.

**Reply by wychwitch** [reply]
@poetaman 
> @wychwitch The change needed would be simple. Here are some experiments needed:
> 
> There is a Linux command notify-send clone for WSL here: [wsl-notify-send](https://github.com/stuartleeks/wsl-notify-send), and a related [article](https://stuartleeks.com/posts/wsl-github-cli-windows-notifications-part-1/). The first thing to do is check if that really works. Also, does it make some default sound? Please try the following commands, and post a screenshot of notification.

This is so funny, I had installed wsl-notify-send a few days ago!! 

> ```
> # First add notify-send function to bash or zsh as described on README.md of wsl-notify-send
> notify-send -u critical "ARTTIME" "Sat May 06, 2023 12:16:55AM PDT\rgoal: 1m\rart: buttefly\rID: 23342"
> # Try it without the wrapper too
> wsl-notify-send.exe --category $WSL_DISTRO_NAME -u critical "ARTTIME" "Sat May 06, 2023 12:16:55AM PDT\rgoal: 1m\rart: buttefly\rID: 23342"
> ```

Okay so a few things

By default that command doesn't work, when I run it (either as the notify-send function or the straight .exe) it just spits the help command at me. I realize it's because of how "ARTTIME" is a separate string and its expecting only one

I adjusted it to the following so it replaces the WSL distro text with ARTTIME instead, also removed the urgency since WSL ignores it anyways

```bash
wsl-notify-send.exe --category "ARTTIME" "Sat May 06, 2023 12:16:55AM PDT\rgoal: 1m\rart: buttefly\rID: 23342"
```

However, it doesn't display the newlines correctly:
![image](https://user-images.githubusercontent.com/1291820/236641545-6851bda0-75e8-4b18-8a52-a08c6203ae00.png)
Replacing them with \n's instead of \r doesn't fix it either. However it does work! The sound is the default windows notification sound too.

> If that works, then the next step would be for arttime to detect it is running in WSL. Can you provide the output of these commands: `$ uname`, `$ uname -a`, `uname -s`, and `uname -r`?

uname & uname -s:
`Linux`
uname -a:
`Linux <HOSTNAME> 5.15.90.1-microsoft-standard-WSL2 #1 SMP Fri Jan 27 02:56:13 UTC 2023 x86_64 GNU/Linux`
uname -r (looks like the best one!)
`5.15.90.1-microsoft-standard-WSL2`
 
> An alternative to wsl-notify-send is [BurntToast](https://github.com/Windos/BurntToast). Though it's not necessary to test this if wsl-notify-send works. It requires powershell, so the question then would be... is powershell.exe available from Linux running under WSL by default or is it something one needs to install from [AUR](https://aur.archlinux.org/packages/powershell)?

yes, powershell.exe is available (actually, anything in the windows path is available! I can launch explorer.exe to a specific folder or even open a file from WSL using a windows exe for example!)

It took a while but I finally got BurntToast to run from WLS, but I had to be a little hacky.

First I had to make a script (as I couldn't for the life find a way to do this otherwise) with the following:
```powershell
New-BurntToastNotification -Text "Don't forget to smile!",'Your script ran successfully, celebrate!'
```
I then launch a new powershell window in the bg to run the command and quit:
```bash
powershell.exe Start-Process pwsh "$(wslpath -w .)\test.ps1"
```
Once all that's done, I get the notification!
![image](https://user-images.githubusercontent.com/1291820/236643659-bf3d30d5-cba6-43ec-93f0-6a1c4696566c.png)
I dunno how feasible it is to use BurntToast, but I wanted to try it out a bit anyways since it looks much more powerful.

> zsh is pretty compatible with bash syntax for basic operations. But I can do the edits in a branch for you to test based on the results of experiments mentioned above.

I would appreciate that!! Only if you have time to ^^

> Thanks for your willingness and offer to collaborate.

Of course!!!! I hope this was hopeful

**Reply by poetaman** [reply]
@wychwitch Looks great! I created an issue https://github.com/poetaman/arttime/issues/53 that refers this testing. It would be good to follow up there as other contributors on this discussion might be receiving notifications they didn't expect. Time is relative. I would like this feature in, though might be some time before I get a chance. Have left you a followup test in that issue :)

**Reply by poetaman** [reply]
@panki27 From https://github.com/poetaman/arttime/issues/53, it looks like arttime now works well under WSL with native Windows notifications and sounds. The dependencies are: ``zsh`` (>=5.7), [wsl-notify-send](https://github.com/stuartleeks/wsl-notify-send), and ``less``. Can you please pull the main branch, ``cd`` to arttime repo, and try ``TERM=xterm-256color ./bin/arttime`` after installing the dependencies? If it doesn't work, or you have some issues, feel free to drop a comment here or open a new issue.

**Reply by panki27** [reply]
@poetaman I got around to testing today. I set a goal `30s` and got a Windows popup at the correct time 👍 
Anything else you want me to try?

Now if only the Conpty issue were fixed already 😞

**Reply by poetaman** [reply]
@panki27 Thanks! Conpty issue affects only a known subset of art in arttime and it cannot do anything about that. arttime seems feature-complete for WSL. Maybe we just need to make more noise on the Microsoft GitHub page.

**Reply by panki27** [reply]
@poetaman It's been a while, but lo and behold:

<img width="410" height="643" alt="grafik" src="https://github.com/user-attachments/assets/ea2f0016-1a63-42dd-998a-36125d418979" />

ConPTY issue seems fixed! 🥳 @wychwitch can you confirm?

**Reply by wychwitch** [reply]
OH wow it has been a long time! Ive since changed my setup drastically so it would take a while to test it. If anyone else is able to beat me to it feel free! I'll try to test it before the end of the week ^^
